### PR TITLE
Sample "isotropic box"

### PR DESCRIPTION
**Pull request by yxqd**
This is first attempt to implement a sample component. This is not a generic component that supports various shapes and kernels, but a simple one that has a box shape and an isotropic kernel.

### CODE REVIEW

**Comment by yxqd** [comment]
@mtbc For testing the new component, I refactored the `instrument_factory` module in `tests/components/optics/`. The majority of your code for instrument_factory was moved to the `mcvine.acc.test.instrument_factory` module, but some code specific for testing guides are left in `tests/components/optics/`. Please check.

**Comment by yxqd** [comment]
@mtbc I also moved the test_helper module to `mcvine.acc.test.compare_acc_nonacc` and made it a bit more customizable

**Comment by yxqd** [comment]
@mtbc Could you please check the revisions to [mcvine.acc.test](https://github.com/mcvine/acc/tree/isotropic_box/acc/test) and tests in [tests/components/optics](https://github.com/mcvine/acc/tree/isotropic_box/tests/components/optics)? Thanks

File: acc/components/samples/isotropic_box.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        NB_FLOAT[:],
+        NB_FLOAT[:]
+    ), device=True)
+    def propagate(
+            threadindex, rng_states,
+            neutron, param_arr
+    ):
+        xwidth, yheight, zthickness, mu, sigma = param_arr
+        x, y, z, vx, vy, vz = neutron[:6]
+        t1, t2 = cu_device_intersect_box(x,y,z, vx,vy,vz, xwidth, yheight, zthickness)
```
Comment (mtbc) [inline acc/components/samples/isotropic_box.py:63]: Just want to check, what does it return if it never intersects?

File: acc/components/samples/isotropic_box.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        name (str): the name of this component
+        xwidth (m): width
+        yheight (m): height
+        zthickness (m): thickness
+        mu (1/m): inverse absorption length
+        sigma (1/m): inverse scattering length
+        """
+        self.name = name
+
+        # Note the configuration of the slit.
```
Comment (mtbc) [inline acc/components/samples/isotropic_box.py:41]: Does it have a slit?

File: acc/kernels/isotropic.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    sin_t = sqrt(1-cos_t*cos_t)
+    sin_p, cos_p = sin(phi), cos(phi)
+    # compute velocity
+    vx, vy, vz = neutron[3:6]
+    vi = sqrt(vx*vx+vy*vy+vz*vz)
+    vx = vi*sin_t*cos_p
+    vy = vi*sin_t*sin_p
+    vz = vi*cos_t
+    neutron[3:6] = vx,vy,vz
+    return
```
Comment (mtbc) [inline acc/kernels/isotropic.py:21]: It's okay that this one doesn't adjust `neutron[-1]`, I guess the weight/probability?

File: acc/components/samples/isotropic_box.py (hunk 2 above)
Comment (yxqd) [reply acc/components/samples/isotropic_box.py:41]: A sample does not need a slit in it.

File: acc/kernels/isotropic.py (hunk 3 above)
Comment (yxqd) [reply acc/kernels/isotropic.py:21]: Yes for this particular.

File: acc/components/samples/isotropic_box.py (hunk 2 above)
Comment (yxqd) [reply acc/components/samples/isotropic_box.py]: just saw it mentioned in the comment. will remove it

**Pull request merged**